### PR TITLE
rework certificate handling

### PR DESCRIPTION
**Pull request by JesusMcCloud**
* JwsHeader now contains CertificateChain

 * Certificate SANs and IANs can now be parsed.

* Fix RelativeDistinguishedName structure

Tested only using certs from resources (none fail, but they don't cover everything). Still, if it is good enough for Google, it should e good enough for us

### CODE REVIEW

File: datatypes-jws/src/commonMain/kotlin/at/asitplus/crypto/datatypes/jws/JsonWebToken.kt [hunk 1]
```diff
@@ -4,6 +4,7 @@ package at.asitplus.crypto.datatypes.jws
 
 import at.asitplus.KmmResult.Companion.wrap
 import at.asitplus.crypto.datatypes.io.ByteArrayBase64Serializer
+import at.asitplus.crypto.datatypes.jws.io.InstantLongSerializer
```
Comment (nodh) [inline datatypes-jws/src/commonMain/kotlin/at/asitplus/crypto/datatypes/jws/JsonWebToken.kt:7]: Why was this line added, is it probably unused?

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/AlternativeNames.kt [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        }
+
+    val x400Addresses: List<Asn1Sequence> =
+        extensions.filter { it.tag == SubjectAltNameImplicitTags.x400Address }.apply {
+            forEach {
+                if (it !is Asn1Sequence) throw Asn1StructuralException("Invalid x400Address Alternative Name found: ${it.toDerHexString()}")
+            }
+        }.map {
+            (it as Asn1Sequence).also {
+                //TODO: strict structural parsing
```
Comment (nodh) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/AlternativeNames.kt:78]: relevant?

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/AlternativeNames.kt [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            }
+        }.map { ObjectIdentifier.parse((it as Asn1Primitive).content) }
+
+    private fun parseStringSANs(implicitTag: UByte) =
+        extensions.filter { it.tag == implicitTag }.apply {
+            forEach { if (it !is Asn1Primitive) throw Asn1StructuralException("Invalid dnsName Alternative Name found: ${it.toDerHexString()}") }
+        }.map { (it as Asn1Primitive).content.decodeToString() }
+
+    override fun toString(): String {
+        //StringBuilder()
```
Comment (nodh) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/AlternativeNames.kt:95]: commented out code

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/AlternativeNames.kt [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            forEach {
+                if (it !is Asn1Sequence) throw Asn1StructuralException("Invalid ediPartyName Alternative Name found: ${it.toDerHexString()}")
+            }
+        }.map {
+            (it as Asn1Sequence).also {
+                if (it.children.size > 2) throw Asn1StructuralException("Invalid partyName Alternative Name found (>2 children): ${it.toDerHexString()}")
+                if (it.children.find { it.tag != 0u.toImplicitTag() && it.tag != 1u.toImplicitTag() } != null) throw Asn1StructuralException(
+                    "Invalid partyName Alternative Name found (illegal implicit tag): ${it.toDerHexString()}"
+                )
+                //TODO: strict string parsing
```
Comment (nodh) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/AlternativeNames.kt:67]: revelant todo?

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/RelativeDistinguishedName.kt [hunk 5]
```diff
@@ -1,37 +1,45 @@
...
+        override fun decodeFromTlv(src: Asn1Set): RelativeDistinguishedName = runRethrowing {
+            RelativeDistinguishedName(src.children.map { AttributeTypeAndValue.decodeFromTlv(it as Asn1Sequence) })
+        }
+    }
+
+    override fun toString() = "DistinguishedName(attrsAndValues=${attrsAndValues.joinToString()})"
+
+}
 
 //TODO: value should be Asn1Primitive???
```
Comment (nodh) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/RelativeDistinguishedName.kt:33]: is this todo still relevant?

**Review by nodh — APPROVED** [review]
Just some minor remarks

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/AlternativeNames.kt (hunk 4 above)
Comment (JesusMcCloud) [reply datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/AlternativeNames.kt:67]: that is a larger TODO. we are simply a bit too lenient (As we've always been)

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/AlternativeNames.kt (hunk 2 above)
Comment (JesusMcCloud) [reply datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/AlternativeNames.kt:78]: that is a larger TODO. we are simply a bit too lenient (As we've always been)

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/RelativeDistinguishedName.kt (hunk 5 above)
Comment (JesusMcCloud) [reply datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/RelativeDistinguishedName.kt:33]: someone would need to study the RFC ;-)

File: datatypes-jws/src/commonMain/kotlin/at/asitplus/crypto/datatypes/jws/JsonWebToken.kt (hunk 1 above)
Comment (JesusMcCloud) [reply datatypes-jws/src/commonMain/kotlin/at/asitplus/crypto/datatypes/jws/JsonWebToken.kt:7]: the file moved and so did the package declaration

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/AlternativeNames.kt (hunk 3 above)
Comment (JesusMcCloud) [reply datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/AlternativeNames.kt]: thanks, removed!
